types/rlp_unmarshal_storage.go

```diff
@@ -115,30 +115,40 @@ func (r *Receipt) UnmarshalStoreRLPFrom(p *fastrlp.Parser, v *fastrlp.Value) err
 		return err
 	}
 
-	if len(elems) != 3 {
-		return fmt.Errorf("expected 3 elements")
+	if len(elems) < 3 {
+		return fmt.Errorf("expected at least 3 elements")
```

Comment:
@lazartravica @Kourin1996 @dbrajovic 

This is a breaking change - please note that the conditional here was `!=` and not `<`.
This means that whatever you add to the `Receipts` struct additionally during marshalling now **will not get unmarshalled** on older versions of the codebase (clients) - because they expect _exactly_ 3 elements.

This unmarshalling problem is present with almost all RLP structures, in a sense that there is room for basically 0 backwards compatibility. I am facing this problem with my block validation task (I wanted to add receipts, but it broke the code for older clients).

We need to decide on how we want to treat these problems going forward:
* make this change, release a version that's a major update (non-compatible with older versions of the code). It essentially means that partners would need to update **all** nodes to the latest version in order for everything to work
* continue to write 600 lines of code for backwards compatibility for every small fix or feature we want to add, because the base architecture is wrong and needs to be changed at the core

Reply:
I think that our best option is to release this as a major change, but before release create PR which fixes all similar issues in the package. It will be a few days of work but will make our life easier in the long run.
Adding additional logic will just make a mess in our code which will catch up with us later

Reply:
Resolving as we continued the discussion internally